src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -80,24 +70,35 @@ internal enum PreconditionState
             }
 
             var serveBody = !HttpMethods.IsHead(request.Method);
-            if (HttpMethods.IsHead(request.Method) || HttpMethods.IsGet(request.Method))
+            var preconditionState = GetPreconditionState(context, httpRequestHeaders, lastModified, etag);
+            if (preconditionState == PreconditionState.NotModified)
             {
-                var preconditionState = GetPreconditionState(context, httpRequestHeaders, lastModified, etag);
-                if (request.Headers.ContainsKey(HeaderNames.Range) &&
-                    (preconditionState == PreconditionState.Unspecified ||
-                    preconditionState == PreconditionState.ShouldProcess))
-                {
-                    return SetRangeHeaders(context, httpRequestHeaders, fileLength, lastModified, etag);
-                }
-                if (preconditionState == PreconditionState.NotModified)
-                {
-                    serveBody = false;
-                    response.StatusCode = StatusCodes.Status304NotModified;
-                }
-                else if (preconditionState == PreconditionState.PreconditionFailed)
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status304NotModified;
+            }
+            else if (preconditionState == PreconditionState.PreconditionFailed)
+            {
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status412PreconditionFailed;
+            }
+
+            if (fileLength.HasValue)
+            {
+                SetAcceptRangeHeader(context);
+                // Assuming the request is not a range request, the Content-Length header is set to the length of the entire file. 
+                // If the request is a valid range request, this header is overwritten with the length of the range as part of the 
+                // range processing (see method SetContentLength).
+                response.ContentLength = fileLength.Value;
+                if (HttpMethods.IsHead(request.Method) || HttpMethods.IsGet(request.Method))
                 {
-                    serveBody = false;
-                    response.StatusCode = StatusCodes.Status412PreconditionFailed;
+                    if ((preconditionState == PreconditionState.Unspecified ||
```

Comment:
If you add the returns above I don't think you need to check preconditionState again here.